var documenterSearchIndex = {"docs":
[{"location":"population/#Population","page":"Population","title":"Population","text":"","category":"section"},{"location":"population/","page":"Population","title":"Population","text":"Types and functions used for managing population data (locations, genotypes, and growth rates).","category":"page"},{"location":"population/#Types","page":"Population","title":"Types","text":"","category":"section"},{"location":"population/","page":"Population","title":"Population","text":"HZAM.Population.Location\nHZAM.Population.Zone\nHZAM.Population.PopulationData","category":"page"},{"location":"population/#HZAM.Population.Location","page":"Population","title":"HZAM.Population.Location","text":"Location\n\nA Location stores an x coordinate and a y coordinate.\n\nConstructors\n\n- Location(x::Float32, y::Float32)\n- Location(starting_range::Vector{<:Location})\n- Location(starting_location::Location, sigma_disp::Real)\n\nDetails on behaviour of different constructors\n\nThe first constructor creates a new location with the given x and y coordinates.\n\nThe second constructor generates a random location within the given range. Range is  specified by a vector containing the coordinates of the bottom left and top right corners.\n\nThe third constructor generates an offspring's location based on the mother's location and  the standard deviation of the dispersal distance.\n\n\n\n\n\n","category":"type"},{"location":"population/#HZAM.Population.Zone","page":"Population","title":"HZAM.Population.Zone","text":"Zone\n\nA Zone stores the genotypes locations and resource use for the individuals contained within  a subset of the range.\n\nFields\n\ngenotypes_F::Vector{Matrix{Int8}}: the female genotypes. rows are alleles (row 1 from mother, row 2 from father) and columns are loci.\ngenotypes_M::Vector{Matrix{Int8}}: the male genotypes.\nlocations_F::Vector{<:Location}: the female locations.\nlocations_M::Vector{<:Location}: the male locations.\nind_useResourceA_F::Vector{Float64}: each female's contribution to the use of resource A.\nind_useResourceA_M::Vector{Float64}: each male's contribution to the use of resource A.\nind_useResourceB_F::Vector{Float64}: each female's contribution to the use of resource B.\nind_useResourceB_M::Vector{Float64}: each male's contribution to the use of resource B.\n\nConstructors\n\n- Zone(\n    starting_N::Integer,\n    total_loci::Integer,\n    location::Location,\n    size::Float32,\n    population::Real,\n    ecolDiff::Real\n)\n- Zone(\n    genotypes_F::Vector{Matrix{Int8}},\n    genotypes_M::Vector{Matrix{Int8}},\n    locations_F::Vector{<:Location},\n    locations_M::Vector{<:Location},\n    competition_trait_loci::Union{UnitRange{<:Integer},Vector{<:Integer}},\n    ecolDiff\n)\n\nDetails on behaviour of different constructors\n\nThe first constructor sets up the initial locations, genotypes, and resource use of every  individual in the zone.\n\nThe second constructor creates a new Zone using the genotypes and locations of the  offspring.\n\n\n\n\n\n","category":"type"},{"location":"population/#HZAM.Population.PopulationData","page":"Population","title":"HZAM.Population.PopulationData","text":"PopulationData\n\nA PopulationData stores the genotypes, locations, and growth rates for all individuals in  the simulation. The data is subdivided by 'zone' into a matrix for  calculation efficiency.\n\nFields\n\npopulation::Matrix{Zone}: all of the zones (containing the genotypes, locations, and resource use) in the simulation.\ngrowth_rates_F::Matrix{Vector{Float64}}: the growth rates of every female for each zone.\n\nConstructors\n\n- PopulationData(\n    K_total::Integer,\n    ecolDiff::Real,\n    total_loci::Integer,\n    intrinsic_R::Real,\n    sigma_comp::Real\n)\n- PopulationData(\n    genotypes_daughters::Matrix{Vector{Matrix{Int8}}},\n    genotypes_sons::Matrix{Vector{Matrix{Int8}}},\n    locations_daughters::Matrix{Vector{Location}},\n    locations_sons::Matrix{Vector{Location}},\n    competition_trait_loci::Union{UnitRange{<:Integer},Vector{<:Integer}},\n    K_total::Integer,\n    sigma_comp::Real,\n    intrinsic_R::Real,\n    ecolDiff::Real\n)\n\nDetails on behaviour of different constructors\n\nThe first constructor sets up the initial population for the simulation.\n\nThe second constructor creates a new PopulationData using the genotypes and locations of the  offspring.\n\n\n\n\n\n","category":"type"},{"location":"population/#Functions","page":"Population","title":"Functions","text":"","category":"section"},{"location":"population/","page":"Population","title":"Population","text":"HZAM.Population.assign_zone(location::HZAM.Population.Location)\nHZAM.Population.calc_ind_useResourceA(competition_traits::Vector, ecolDiff::Real)\nHZAM.Population.calc_ind_useResourceB(competition_traits::Vector, ecolDiff::Real)\nHZAM.Population.max_radius_squared(location::HZAM.Population.Location, t::Real, max_dist::Real)\nHZAM.Population.calc_ideal_densities(\n    K_total::Integer,\n    sigma_comp::Real,\n    locations_F::Vector{<:HZAM.Population.Location},\n    max_dist::Real\n)\nHZAM.Population.calc_squared_distances(location_list::Vector{<:HZAM.Population.Location}, focal_location::HZAM.Population.Location)\nHZAM.Population.get_surrounding_zones(zone_index::CartesianIndex)\nHZAM.Population.calc_real_densities(\n    neighbourhood::Matrix{HZAM.Population.Zone},\n    locations_F::Vector{<:HZAM.Population.Location},\n    max_dist::Real,\n    sigma_comp::Real\n)\nHZAM.Population.calc_growth_rates(\n    population::Matrix{HZAM.Population.Zone},\n    zone_index::CartesianIndex,\n    K_total::Integer,\n    sigma_comp::Real,\n    intrinsic_R::Real\n)\nHZAM.Population.genotype_mean(\n    genotype::Matrix{<:Integer},\n    loci::Union{UnitRange{<:Integer},Vector{<:Integer}}\n)\nHZAM.Population.calc_survival_fitness_hetdisadvantage(\n    genotype::Matrix{<:Integer},\n    w_hyb::Real\n)\nHZAM.Population.calc_survival_fitness_epistasis(genotype::Matrix{<:Integer},\n    hybrid_survival_loci::Union{UnitRange{<:Integer},Vector{<:Integer}},\n    w_hyb::Real,\n    beta::Real=1\n)","category":"page"},{"location":"population/#HZAM.Population.assign_zone-Tuple{HZAM.Population.Location}","page":"Population","title":"HZAM.Population.assign_zone","text":"assign_zone(location::Location)\n\nDetermine which zone a location falls in.\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.calc_ind_useResourceA-Tuple{Vector, Real}","page":"Population","title":"HZAM.Population.calc_ind_useResourceA","text":"calc_ind_useResourceA(competition_traits::Vector{Real}, ecolDiff::Real)\n\nCompute the individual contributions to the use of resource A according to a linear gradient  between the resource use of species 0 and species 1.\n\nArguments\n\ncompetition_traits::Vector{Real}: the ecological competition trait values.\necolDiff::Real: the ecological difference between the species. \n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.calc_ind_useResourceB-Tuple{Vector, Real}","page":"Population","title":"HZAM.Population.calc_ind_useResourceB","text":"calc_ind_useResourceB(competition_traits::Vector{Real}, ecolDiff::Real)\n\nCompute the individual contributions to the use of resource B according to a linear gradient  between the resource use of species 0 and species 1.\n\nArguments\n\ncompetition_traits::Vector{Real}: the ecological competition trait values.\necolDiff::Real: the ecological difference between the species. \n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.max_radius_squared-Tuple{HZAM.Population.Location, Real, Real}","page":"Population","title":"HZAM.Population.max_radius_squared","text":"max_radius_squared(x::Real, y::Real, t::Real, max_dist::Real)\n\nCompute the distance from a point along an angle to the limit of the range  (defined by x∈[0,1), y∈[0,1)). \n\nThe distance gets cut off at max_dist. Used in calculating the ideal densities.\n\nArguments\n\nlocation::Location: the starting location.\nt::Real: the angle.\nmax_dist: the maximum distance.\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.calc_ideal_densities-Tuple{Integer, Real, Vector{<:HZAM.Population.Location}, Real}","page":"Population","title":"HZAM.Population.calc_ideal_densities","text":"calc_ideal_densities(\n    K_total::Integer,\n    sigma_comp::Real,\n    locations_F::Vector{<:Location},\n    max_dist::Real\n)\n\nCompute expected local densities, based on geographically even distribution of individuals  at carrying capacity.\n\nDensities are calculated using the following integral:\n\nK_totalintlimits_0^2piintlimits_0^003  exp(-fracr^22σ_comp^2)rdrdθ\n\nwhere the density function is equal to 0 outside of the range limits.\n\nArguments\n\nK_total::Integer: the carrying capacity.\nsigma_comp::Real: the standard deviation for the normal curve used in calculating local density.\nlocations_F::Vector{<:Location}: a list of the locations where the ideal density is to be computed.\nmax_dist::Real: the cutoff for the furthest away an individual can be and affect the density calculation. Should be 3x sigma_comp.\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.calc_squared_distances-Tuple{Vector{<:HZAM.Population.Location}, HZAM.Population.Location}","page":"Population","title":"HZAM.Population.calc_squared_distances","text":"calc_squared_distances(location_list::Vector{<:Location}, focal_location::Location)\n\nCompute the squared distances from a set of locations to a single location.\n\nExample\n\njulia> calc_squared_distances([Location(0.5f0, 0.5f0), Location(0.4f0, 0.4f0), \nLocation(0f0,0f0)], Location(0.5f0, 0.5f0))\n3-element Vector{Float32}:\n 0.0\n 0.019999998\n 0.5\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.get_surrounding_zones-Tuple{CartesianIndex}","page":"Population","title":"HZAM.Population.get_surrounding_zones","text":"get_surrounding_zones(zone_index::CartesianIndex)\n\nReturn a list of the indices of the neighbouring zones and the given zone index.\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.calc_real_densities-Tuple{Matrix{HZAM.Population.Zone}, Vector{<:HZAM.Population.Location}, Real, Real}","page":"Population","title":"HZAM.Population.calc_real_densities","text":"calc_real_densities(\n    neighbourhood::Matrix{Zone},\n    locations_F::Vector{<:Location},\n    max_dist::Real,\n    sigma_comp::Real\n)\n\nCompute the population density at each female's location for both resources.\n\nArguments\n\nneighbourhood::Matrix{Zone}: the zone indices for all the zones close enough to affect the density calculation.\nlocations_F::Vector{<:Location}: the locations of every female in the focal zone.\nmax_dist::Real: the distance cutoff for the density calculation.\nsigma_comp::Real: the standard deviation used in the density calculation.\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.calc_growth_rates-Tuple{Matrix{HZAM.Population.Zone}, CartesianIndex, Integer, Real, Real}","page":"Population","title":"HZAM.Population.calc_growth_rates","text":"calc_growth_rates(\n    population::Matrix{Zone},\n    zone_index::CartesianIndex,\n    K_total::Integer,\n    sigma_comp::Real,\n    intrinsic_R::Real\n)\n\nCompute the female growth rates.\n\nArguments\n\npopulation::Matrix{Zone}: the matrix of zones storing all the locations, and genotypes in the simulation.\nzone_index::CartesianIndex: the index of the focal zone.\nK_total::Integer: the carrying capacity.\nsigma_comp::Real: the standard deviation for the normal curve used in calculating local density.\nintrinsic_R::Real: the intrinsic growth rate.\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.genotype_mean-Tuple{Matrix{<:Integer}, Union{UnitRange{<:Integer}, Vector{<:Integer}}}","page":"Population","title":"HZAM.Population.genotype_mean","text":"genotype_mean(genotype::Matrix{<:Integer}, loci::Union{UnitRange{<:Integer},Vector{<:Integer}})\n\nCompute the mean value of the genotype across the list of loci given.\n\nExample\n\njulia> genotype_mean([0 1 0; 0 1 0], [1,3])\n0.0\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.calc_survival_fitness_hetdisadvantage-Tuple{Matrix{<:Integer}, Real}","page":"Population","title":"HZAM.Population.calc_survival_fitness_hetdisadvantage","text":"calc_survival_fitness_hetdisadvantage(genotype::Matrix{<:Integer}, w_hyb::Real)\n\nCompute the survival fitness of an individual according to heterozygosity.\n\nArguments\n\ngenotype:Matrix{<:Integer}: the individual's genotype.\nw_hyb::Real: the simulation's hybrid fitness value.\n\n\n\n\n\n","category":"method"},{"location":"population/#HZAM.Population.calc_survival_fitness_epistasis","page":"Population","title":"HZAM.Population.calc_survival_fitness_epistasis","text":"calc_survival_fitness_epistasis(\n    genotype::Matrix{<:Integer}, \n    hybrid_survival_loci::Union{UnitRange{<:Integer},Vector{<:Integer}}, \n    w_hyb::Real, \n    beta::Real=1\n)\n\nCompute the survival fitness of an individual according to epistasis, with the beta  parameter set to one as a default.\n\nArguments\n\ngenotype:Matrix{<:Integer}: the individual's genotype.\nhybrid_survival_loci::Union{UnitRange{<:Integer},Vector{<:Integer}}: the loci responsible for heterozygote disadvantage in computing the survival fitness.\nw_hyb::Real: the simulation's hybrid fitness value.\nbeta::Real\n\n\n\n\n\n","category":"function"},{"location":"mating/#Mating","page":"Mating","title":"Mating","text":"","category":"section"},{"location":"mating/","page":"Mating","title":"Mating","text":"Functions involved in mate selection and generating the offspring genotype.","category":"page"},{"location":"mating/#Functions","page":"Mating","title":"Functions","text":"","category":"section"},{"location":"mating/","page":"Mating","title":"Mating","text":"HZAM.Mating.distance(location1::HZAM.Population.Location, location2::HZAM.Population.Location)\nHZAM.Mating.choose_closest_male(\n        zones::Matrix{HZAM.Population.Zone}, \n        zone_indices::Vector{CartesianIndex{2}}, \n        elig_M::Dict{CartesianIndex,<:Vector{<:Integer}}, \n        location_mother::HZAM.Population.Location, \n        neighbourhood_size::Float32\n    )\nHZAM.Mating.choose_closest_male_from_zone(\n        elig_M::Vector{<:Integer}, \n        locations_M::Vector{HZAM.Population.Location}, \n        location_mother::HZAM.Population.Location\n    )\nHZAM.Mating.calc_match_strength(\n        female_genotype::Matrix{<:Integer}, \n        male_genotype::Matrix{<:Integer}, \n        pref_SD::Float64, \n        female_mating_trait_loci, \n        male_mating_trait_loci\n    )\nHZAM.Mating.generate_offspring_genotype(\n        mother_genotype::Matrix{<:Integer}, \n        father_genotype::Matrix{<:Integer}\n    )","category":"page"},{"location":"mating/#HZAM.Mating.distance-Tuple{HZAM.Population.Location, HZAM.Population.Location}","page":"Mating","title":"HZAM.Mating.distance","text":"distance(location1::Location, location2::Location)\n\nCompute the distance between location1 and location2.\n\n\n\n\n\n","category":"method"},{"location":"mating/#HZAM.Mating.choose_closest_male-Tuple{Matrix{HZAM.Population.Zone}, Vector{CartesianIndex{2}}, Dict{CartesianIndex, <:Vector{<:Integer}}, HZAM.Population.Location, Float32}","page":"Mating","title":"HZAM.Mating.choose_closest_male","text":"choose_closest_male(\n    zones::Matrix{Zone}, \n    zone_indices::Vector{CartesianIndex{2}}, \n    elig_M::Dict{CartesianIndex,<:Vector{<:Integer}}, \n    location_mother::Location, \n    neighbourhood_size::Float32\n)\n\nFind the index of the closest male within an area covering multiple zones from a list of  eligible males. \n\nReturn both the male index and its zone index. Return -1 for both values if  no eligible male is found.\n\nArguments\n\nzones::Matrix{Zone}: the matrix storing the population data for each zone.\nzone_indices::Vector{CartesianIndex{2}}: the indices of the zones to be checked for the closest male.\nelig_M::Dict{CartesianIndex,<:Vector{<:Integer}}: the indices of all the eligible males in each zone.\nlocation_mother::Location: the location from which the males' distances are computed.\nneighbourhood_size::Float32: the distance cutoff for the search.\n\n\n\n\n\n","category":"method"},{"location":"mating/#HZAM.Mating.choose_closest_male_from_zone-Tuple{Vector{<:Integer}, Vector{HZAM.Population.Location}, HZAM.Population.Location}","page":"Mating","title":"HZAM.Mating.choose_closest_male_from_zone","text":"choose_closest_male_from_zone(\n    elig_M::Vector{<:Integer}, \n    locations_M::Vector{Location}, \n    location_mother::Location\n)\n\nFind the index of the closest male within a single zone from a list of eligible males.\n\nArguments\n\nelig_M::Vector{<:Integer}: the indices of all the eligible males.\nlocations_M::Vector{Location}: the locations of all the males in the zone.\nlocation_mother::Location: the location from which the males' distances are computed.\n\n\n\n\n\n","category":"method"},{"location":"mating/#HZAM.Mating.calc_match_strength-Tuple{Matrix{<:Integer}, Matrix{<:Integer}, Float64, Any, Any}","page":"Mating","title":"HZAM.Mating.calc_match_strength","text":"calc_match_strength(\n    female_genotype::Matrix{<:Integer}, \n    male_genotype::Matrix{<:Integer}, \n    pref_SD::Float64, \n    female_mating_trait_loci, \n    male_mating_trait_loci\n)\n\nCompare the male's mating trait with the female mating trait to determine the match strength  along a Gaussian acceptance curve. \n\nThe acceptance curve is bounded between 0 and 1.\n\nArguments\n\nfemale_genotype::Matrix{<:Integer}: the genotype of the focal female.\nmale_genotype::Matrix{<:Integer}: the genotype of the focal male.\npref_SD::Float64: width of the Gaussian acceptance curve.\nfemale_mating_trait_loci: the loci contributing to the female's mate preference.\nmale_mating_trait_loci: the loci contributing to the male's trait.\n\n\n\n\n\n","category":"method"},{"location":"mating/#HZAM.Mating.generate_offspring_genotype-Tuple{Matrix{<:Integer}, Matrix{<:Integer}}","page":"Mating","title":"HZAM.Mating.generate_offspring_genotype","text":"generate_offspring_genotype(\n    mother_genotype::Matrix{<:Integer}, \n    father_genotype::Matrix{<:Integer}\n)::Matrix{<:Integer}\n\nGenerate the offspring genotype from the parent genotypes. \n\nFor each locus (column), first row for allele from mother, second row for allele from  father. The number of loci in each genotype must be equal.\n\nExample\n\njulia> generate_offspring_genotype([0 0 0; 0 0 0], [1 1 1; 1 1 1])\n2×3 Matrix:\n 0  0  0\n 1  1  1\n\n\n\n\n\n","category":"method"},{"location":"plot_data/#PlotData","page":"PlotData","title":"PlotData","text":"","category":"section"},{"location":"plot_data/","page":"PlotData","title":"PlotData","text":"Functions for plotting data while the simulation is running.","category":"page"},{"location":"plot_data/#Functions","page":"PlotData","title":"Functions","text":"","category":"section"},{"location":"plot_data/","page":"PlotData","title":"PlotData","text":"HZAM.PlotData.create_population_plot(\n    hybrid_indices_functional::Vector,\n    locations::Vector,\n    save_plot::Bool\n)\nHZAM.PlotData.update_population_plot(\n    hybrid_indices_functional::Vector,\n    locations::Vector,\n    generation::Integer,\n    save_plot::Bool\n)\nHZAM.PlotData.scale_curve(curve, Index::Integer)\nHZAM.PlotData.count_genotypes(\n        range::Union{UnitRange{<:Integer},Vector{<:Integer}},\n        genotypes::Vector{<:Matrix{<:Integer}}\n    )\nHZAM.PlotData.get_expected(x::Real, n::Integer)\nHZAM.PlotData.create_gene_plot(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::NamedTuple,\n    save_plot::Bool\n)\nHZAM.PlotData.update_gene_plot(\n        genotypes::Vector{<:Matrix{<:Integer}},\n        loci::NamedTuple,\n        generation::Integer,\n        save_plot::Bool\n    )","category":"page"},{"location":"plot_data/#HZAM.PlotData.create_population_plot-Tuple{Vector, Vector, Bool}","page":"PlotData","title":"HZAM.PlotData.create_population_plot","text":"create_population_plot(\n    hybrid_indices_functional::Vector{<:Real},\n    locations::Vector,\n    save_plot::Bool\n)\n\nInitialize the plot of locations and hybrid indices.\n\nArguments\n\nhybrid_indices_functional::Vector: list of the hybrid index (value between 0 and 1) of \n\nevery individual.\n\nlocations::Vector: the location of every individual \n\n(must be in the same order as the hybrid indices).\n\ngeneration::Integer: the number of elapsed generations.\nsave_plot::Bool: true if the plot is to be saved to a PNG file.\n\n\n\n\n\n","category":"method"},{"location":"plot_data/#HZAM.PlotData.update_population_plot-Tuple{Vector, Vector, Integer, Bool}","page":"PlotData","title":"HZAM.PlotData.update_population_plot","text":"update_population_plot(\n    hybrid_indices_functional::Vector,\n    locations::Vector,\n    generation::Integer,\n    save_plot::Bool\n)\n\nUpdate the existing plot with new locations and hybrid indices.\n\nArguments\n\nhybrid_indices_functional::Vector: list of the hybrid index (value between 0 and 1) of \n\nevery individual.\n\nlocations::Vector: the location of every individual (must be in the same order as the \n\nhybrid indices).\n\ngeneration::Integer: the number of elapsed generations.\nsave_plot::Bool: true if the plot is to be saved to a PNG file.\n\n\n\n\n\n","category":"method"},{"location":"plot_data/#HZAM.PlotData.scale_curve-Tuple{Any, Integer}","page":"PlotData","title":"HZAM.PlotData.scale_curve","text":"scale_curve(curve, index::Integer)\n\nScale the curve output to display on the plot over the corresponding y-axis range.\n\nArguments\n\ncurve: the output of a function approximating the data whose range is [0,1].\nindex::Integer: the index of the corresponding range on the y-axis. 1 refers to \n\n[0, 0.1), 2 to [0.1, 0.2) and so on.\n\n\n\n\n\n","category":"method"},{"location":"plot_data/#HZAM.PlotData.count_genotypes-Tuple{Union{UnitRange{<:Integer}, Vector{<:Integer}}, Vector{<:Matrix{<:Integer}}}","page":"PlotData","title":"HZAM.PlotData.count_genotypes","text":"function count_genotypes(\n    range::Union{UnitRange{<:Integer},Vector{<:Integer}},\n    genotypes::Vector{<:Matrix{<:Integer}}\n)\n\nCount the number of individuals with each genotype for a given loci range.\n\n\n\n\n\n","category":"method"},{"location":"plot_data/#HZAM.PlotData.get_expected-Tuple{Real, Integer}","page":"PlotData","title":"HZAM.PlotData.get_expected","text":"get_expected(x::Real, n::Integer)\n\nCompute the expected number of individuals with a given trait hybrid index if the trait is  controlled by 4 loci. \n\nArguments\n\nx::Real: the hybrid index for the trait of interest.\nn::Integer: the total population size.\n\n\n\n\n\n","category":"method"},{"location":"plot_data/#HZAM.PlotData.create_gene_plot-Tuple{Vector{<:Matrix{<:Integer}}, NamedTuple, Bool}","page":"PlotData","title":"HZAM.PlotData.create_gene_plot","text":"create_gene_plot(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::NamedTuple,\n    save_plot::Bool\n)\n\nInitialize the plot of the phenotype frequencies for each loci range.\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotypes of every individual.\nloci::NamedTuple: the loci range for each trait.\nsave_plot::Bool: true if the plot is to be saved to a file.\n\n\n\n\n\n","category":"method"},{"location":"plot_data/#HZAM.PlotData.update_gene_plot-Tuple{Vector{<:Matrix{<:Integer}}, NamedTuple, Integer, Bool}","page":"PlotData","title":"HZAM.PlotData.update_gene_plot","text":"update_gene_plot(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::NamedTuple,\n    generation::Integer,\n    save_plot::Bool\n)\n\nUpdate the plot of the phenotype frequencies for each loci range.\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotypes of every individual.\nloci::NamedTuple: the loci range for each trait.\ngeneration::Integer: the generation # the simulation is on.\nsave_plot::Bool: true if the plot is to be saved to a file.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#DataAnalysis","page":"DataAnalysis","title":"DataAnalysis","text":"","category":"section"},{"location":"data_analysis/","page":"DataAnalysis","title":"DataAnalysis","text":"Types and functions used for analyzing data produced by the simulation.","category":"page"},{"location":"data_analysis/#Types","page":"DataAnalysis","title":"Types","text":"","category":"section"},{"location":"data_analysis/","page":"DataAnalysis","title":"DataAnalysis","text":"HZAM.DataAnalysis.SimParams\nHZAM.DataAnalysis.PopulationTrackingData\nHZAM.DataAnalysis.SpatialData\nHZAM.DataAnalysis.OutputData","category":"page"},{"location":"data_analysis/#HZAM.DataAnalysis.SimParams","page":"DataAnalysis","title":"HZAM.DataAnalysis.SimParams","text":"SimParams\n\nA SimParams stores the parameters used in a simulation.\n\nFields\n\nintrinsic_R::Real: the intrinsic growth rate.\necolDiff::Real: the ecological difference between the two populations.\nw_hyb::Real: the hybrid fitness.\nS_AM::Real: the strength of assortative mating.\nK_total::Integer: the carrying capacity of the environment.\nmax_generations::Integer: the number of generations the simulation ran for.\nsigma_disp::Real: the standard deviation for dispersal distance.\ntotal_loci::Integer: the number of loci in the genotypes.\nfemale_mating_trait_loci: the loci controlling mating preference.\nmale_mating_trait_loci: the loci controlling mating cue.\ncompetition_trait_loci: the loci controlling what resources the individual uses.\nhybrid_survival_loci: the loci controlling hybrid fitness.\nper_reject_cost::Real: the loss in fitness incurred by a female after rejecting one male.\n\n\n\n\n\n","category":"type"},{"location":"data_analysis/#HZAM.DataAnalysis.PopulationTrackingData","page":"DataAnalysis","title":"HZAM.DataAnalysis.PopulationTrackingData","text":"PopulationTrackingData\n\nA PopulationTrackingData stores the size, average hybrid index, overlap, and hybrid zone  width of the population.\n\nAs the simulation runs a vector of PopulationTrackingData is used to keep track of the key  population data each generation.\n\nFields\n\npopulation_size::Real: the total number of individuals.\nhybridness::Real: the average hybrid index in the population (0 means phenotypically pure, 1 means that half of the genetic material is from one species and half from the other).\noverlap::Real: the proportion of the range containing both species.\nwidth::Real: the width of the hybrid zone over all functional loci.\n\nConstructors\n\n- PopulationTrackingData(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    locations::Vector,\n    loci::NamedTuple\n)\n\n\n\n\n\n","category":"type"},{"location":"data_analysis/#HZAM.DataAnalysis.SpatialData","page":"DataAnalysis","title":"HZAM.DataAnalysis.SpatialData","text":"SpatialData\n\nA SpatialData stores the bimodality, gene flow, hybrid zone movement, cline width, and cline  position data for the hybrid zone.\n\nFields\n\nbimodality::Real: the proportion of phenotypically pure individuals within half a dispersal distance of the middle of the hybrid zone.\ngene_flows::NamedTuple: the proportion of genetic material originating in the other species found in individuals above a phenotypic purity cutoff.\ncline_widths::NamedTuple: the width of the cline for each trait.\ncline_positions::NamedTuple: the position of the midpoint of the cline for each trait.\nvariance::Real: the average movement of the hybrid zone each generation.\n\nConstructors\n\n- SpatialData(\n        locations::Vector,\n        sigma_disp::Real,\n        genotypes::Vector{<:Matrix{<:Integer}},\n        loci::NamedTuple,\n        last_generation::Bool\n    )\n- SpatialData(output_data::Vector{SpatialData})\n\nDetails on behaviour of different constructors\n\nThe first constructor computes the spatial data from the population's genotypes, standard  deviation in dispersal distance, genotypes, and which loci control which traits.\n\nThe second constructor accepts a vector of SpatialData, computes the variance in the  functional cline position, and the average over time of all remaining fields.\n\n\n\n\n\n","category":"type"},{"location":"data_analysis/#HZAM.DataAnalysis.OutputData","page":"DataAnalysis","title":"HZAM.DataAnalysis.OutputData","text":"OutputData\n\nAn OutputData stores all of the key data from a simulation run.\n\nfields\n\nsim_params::SimParams: the parameters of the simulation.\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotypes of the population at the end of the simulation.\nlocations::Vector{<:Any}: the locations of the population at the end of the simulation.\nspatial_data::Union{SpatialData,Vector{<:SpatialData}}: the spatial characteristics of the hybrid zone over time.\npopulation_tracking_data::Vector{<:PopulationTrackingData}: the key measures of the population over time.\nfitnesses::Vector{<:Dict}: the average number of offspring per phenotype per generation.\nmating_success::Vector{<:Dict}: the average number of mates per male phenotype per generation.\nphenotypes::Vector{NamedTuple}: the number of individuals with each phenotype per generation.\n\n\n\n\n\n","category":"type"},{"location":"data_analysis/#Functions","page":"DataAnalysis","title":"Functions","text":"","category":"section"},{"location":"data_analysis/","page":"DataAnalysis","title":"DataAnalysis","text":"HZAM.DataAnalysis.calc_position(sigmoid_curves::Vector{<:Vector{<:Real}})\nHZAM.DataAnalysis.calc_variance(positions::Vector{<:Real})\nHZAM.DataAnalysis.calc_sigmoid_curve(locations_x::Vector{<:Real}, hybrid_indices::Vector{<:Real})\nHZAM.DataAnalysis.calc_sigmoid_curves(locations::Vector, hybrid_indices::Vector{<:Real})\nHZAM.DataAnalysis.sigmoid(x::Vector{<:Real}, p::Vector{<:Real})\nHZAM.DataAnalysis.calc_width(sigmoid_curve::Vector{<:Real})\nHZAM.DataAnalysis.average_width(sigmoid_curves::Vector{<:Vector{<:Real}})\nHZAM.DataAnalysis.calc_length(sigmoid_curves::Vector{<:Vector{<:Real}})\nHZAM.DataAnalysis.calc_all_gene_flow(\n    genotypes::Vector{<:Matrix{<:Real}},\n    hybrid_indices_functional::Vector{<:Real},\n    loci::NamedTuple\n)\nHZAM.DataAnalysis.calc_all_cline_widths_and_positions(\n        genotypes::Vector{<:Matrix{<:Real}},\n        locations::Vector,\n        loci::NamedTuple\n    )\nHZAM.DataAnalysis.calc_overlap_in_range(\n        locations_x::Vector{<:Real},\n        hybrid_indices_functional::Vector{<:Real}\n    )\nHZAM.DataAnalysis.calc_overlap_overall(\n        locations_x::Vector,\n        hybrid_indices_functional::Vector{<:Real},\n        sorted_indices::Vector{<:Vector{<:Integer}}\n    )\nHZAM.DataAnalysis.calc_bimodality_in_range(\n    sigmoid_curve::Vector{<:Real},\n    locations_x::Vector{<:Real},\n    hybrid_indices::Vector{<:Real},\n    sigma_disp::Real\n)\nHZAM.DataAnalysis.calc_bimodality_overall(\n        sigmoid_curves::Vector{<:Vector{<:Real}},\n        sorted_indices::Vector{<:Vector{<:Integer}},\n        locations_x::Vector{<:Real},\n        hybrid_indices::Vector{<:Real},\n        sigma_disp::Real\n    )\nHZAM.DataAnalysis.calc_linkage_diseq(\n        genotypes::Vector{<:Matrix{<:Integer}},\n        l1::Integer,\n        l2::Integer\n    )\nHZAM.DataAnalysis.calc_average_linkage_diseq(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::NamedTuple\n)\nHZAM.DataAnalysis.calc_trait_correlation(\n        genotypes::Vector{<:Matrix{<:Integer}},\n        loci_range1::Union{UnitRange{<:Integer},Vector{<:Integer}},\n        loci_range2::Union{UnitRange{<:Integer},Vector{<:Integer}}\n    )\nHZAM.DataAnalysis.calc_all_trait_correlations(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::NamedTuple\n)\nHZAM.DataAnalysis.sort_y(y_locations::Vector{<:Real})\nHZAM.DataAnalysis.sort_locations(A::AbstractArray{<:Real}, bin_size::Real)\nHZAM.DataAnalysis.average_gene_data(gene_data::Vector{<:NamedTuple})\nHZAM.DataAnalysis.calc_traits_additive(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::Union{UnitRange{<:Integer},Vector{<:Integer}}\n)\nHZAM.DataAnalysis.average_data_per_phenotype(\n        data::Vector{<:Integer},\n        genotypes::Vector{<:Matrix{<:Integer}},\n        loci::Union{UnitRange{<:Integer},Vector{<:Integer}}\n    )\nHZAM.DataAnalysis.count_phenotypes_at_loci(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::Union{UnitRange{<:Integer},Vector{<:Integer}}\n)\nHZAM.DataAnalysis.find_fixed_alleles(genotypes::Vector{<:Matrix{<:Integer}})\nHZAM.DataAnalysis.calc_chi_squared(genotypes::Vector{<:Matrix{<:Integer}}, locus::Integer)","category":"page"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_position-Tuple{Vector{<:Vector{<:Real}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_position","text":"calc_position(sigmoid_curves::Vector{<:Vector{T} where <:Real})\n\nCalculate the x location at the middle of the cline based on a series of sigmoid curves  fitting the cline at different ranges on the y axis.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_variance-Tuple{Vector{<:Real}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_variance","text":"calc_variance(positions::Vector{<:Real})\n\nCalculate the average movement of the hybrid zone each generation.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_sigmoid_curve-Tuple{Vector{<:Real}, Vector{<:Real}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_sigmoid_curve","text":"calc_sigmoid_curve(locations_x::Vector{<:Real}, hybrid_indices::Vector{<:Real})\n\nCompute a sigmoid curve that models hybrid index vs location on the x axis.\n\nReturn the output of the sigmoid curve function on 1000 evenly spaced locations.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_sigmoid_curves-Tuple{Vector, Vector{<:Real}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_sigmoid_curves","text":"calc_sigmoid_curves(locations::Vector, hybrid_indices::Vector{<:Real})\n\nDivide the range into ten horizontal strips and compute a sigmoid curve fitting the hybrid  indices along each strip.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.sigmoid-Tuple{Vector{<:Real}, Vector{<:Real}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.sigmoid","text":"sigmoid(x::Vector{<:Real}, p::Vector{<:Real})\n\nCompute the y value for a sigmoid given the x value, centre, and maximum slope.\n\nArguments\n\nx::Real: the value along the x axis where the sigmoid is to be calculated.\np::Vector{<:Real}: the fit parameters where p[1] is the cline centre and p[2] is \n\nthe maximum slope.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_width-Tuple{Vector{<:Real}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_width","text":"calc_width(sigmoid_curve::Vector{<:Real})\n\nCompute the width of a cline given a sigmoid curve.\n\nThe cline width is defined as the distance between where the sigmoid curve passes through  0.1 and 0.9.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.average_width-Tuple{Vector{<:Vector{<:Real}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.average_width","text":"average_width(sigmoid_curves::Vector{<:Vector{<:Real}})\n\nCompute the width of a cline given a series of sigmoid curves representing the cline along  horizontal strips of the range.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_length-Tuple{Vector{<:Vector{<:Real}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_length","text":"calc_length(sigmoid_curves::Vector{<:Vector{<:Real}})\n\nCompute the approximate length of the hybrid zone.\n\nThis is done by finding the midpoints of a series of sigmoid curves representing the cline  along evenly spaced horizontal strips of the range. The cline length is defined as the  length of the shortest path traversing the entire range passing through each midpoint.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_all_gene_flow-Tuple{Vector{<:Matrix{<:Real}}, Vector{<:Real}, NamedTuple}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_all_gene_flow","text":"calc_all_gene_flow(\n    genotypes::Vector{<:Matrix{<:Real}},\n    hybrid_indices_functional::Vector{<:Real},\n    loci::NamedTuple\n)\n\nFor each trait, compute the proportion of genetic material originating in the other species  found in individuals above a phenotypic purity cutoff.\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Real}}: the genotypes of all individuals.\nhybrid_indices_functional::Vector{<:Real}: the mean values of the genotypes over the functional loci only.\nloci::NamedTuple: the name and loci range of each trait of interest.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_all_cline_widths_and_positions-Tuple{Vector{<:Matrix{<:Real}}, Vector, NamedTuple}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_all_cline_widths_and_positions","text":"calc_all_cline_widths_and_positions(\n    genotypes::Vector{<:Matrix{<:Real}},\n    locations::Vector,\n    loci::NamedTuple\n)\n\nFor each trait, compute the cline width and the location of the middle.\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Real}}: the genotypes of all individuals.\nlocations::Vector: the locations of all individuals.\nloci::NamedTuple: the name and loci range of each trait of interest.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_overlap_in_range-Tuple{Vector{<:Real}, Vector{<:Real}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_overlap_in_range","text":"calc_overlap_in_range(\n    locations_x::Vector{<:Real},\n    hybrid_indices_functional::Vector{<:Real}\n)\n\nCompute the proportion of the range containing both species.\n\nSubdivide the range into 50 boxes and sums the area covered by boxes containing at least  10% phenotypically pure individuals of both species.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_overlap_overall-Tuple{Vector, Vector{<:Real}, Vector{<:Vector{<:Integer}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_overlap_overall","text":"function calc_overlap_overall(\n    locations_x::Vector,\n    hybrid_indices_functional::Vector{<:Real},\n    sorted_indices::Vector{<:Vector{<:Integer}}\n)\n\nCompute the total overlap area between the two species.\n\nArguments\n\nlocations_x::Vector: the x values of all the locations.\nhybrid_indices_functional::Vector{<:Real}: the mean values of the genotypes over the functional loci only.\nsorted_indices::Vector{<:Vector{<:Integer}}: the indices of the locations sorted into \n\nbins corresponding to non-overlapping ranges of the y values.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_bimodality_in_range-Tuple{Vector{<:Real}, Vector{<:Real}, Vector{<:Real}, Real}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_bimodality_in_range","text":"calc_bimodality_in_range(\n    sigmoid_curve::Vector{<:Real},\n    locations_x::Vector{<:Real},\n    hybrid_indices::Vector{<:Real},\n    sigma_disp::Real\n)\n\nCompute the proportion of phenotypically pure individuals within half a dispersal distance  of the cline midpoint.\n\nArguments\n\nsigmoid_curve::Vector{<:Real}: the sigmoid curve fitting the cline.\nlocations_x::Vector{<:Real}: the x values of the locations.\nhybrid_indices::Vector{<:Real}: the mean values of the genotypes over the functional loci only.\nsigma_disp::Real: the standard deviation in the dispersal distance distribution.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_bimodality_overall-Tuple{Vector{<:Vector{<:Real}}, Vector{<:Vector{<:Integer}}, Vector{<:Real}, Vector{<:Real}, Real}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_bimodality_overall","text":"Compute the bimodality across the entire range (see above for further description).\n\ncalc_bimodality_overall(\n    sigmoid_curves::Vector{<:Vector{<:Real}},\n    sorted_indices::Vector{<:Vector{<:Integer}},\n    locations_x::Vector{<:Real},\n    hybrid_indices::Vector{<:Real},\n    sigma_disp::Real\n)\n\nArguments\n\nsigmoid_curves::Vector{<:Real}: the sigmoid curves for each horizontal strip of the range.\nsorted_indices::Vector{<:Vector{<:Integer}}: the indices of the locations sorted into bins bins corresponding to non-overlapping ranges of the y values.\nlocations_x::Vector{<:Real}: the x values of the locations.\nhybrid_indices::Vector{<:Real}: the mean values of the genotypes over the functional loci only.\nsigma_disp::Real: the standard deviation in the dispersal distance distribution.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_linkage_diseq-Tuple{Vector{<:Matrix{<:Integer}}, Integer, Integer}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_linkage_diseq","text":"calc_linkage_diseq(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    l1::Integer,\n    l2::Integer\n)\n\nCompute the linkage disequilibrium (calculated using the Pearson coefficient) between two  loci.\n\nr = fracsum(x_i -barx)(y_i -bary) sqrtsum(x_i -barx)^2sum(y_i -bary)^2\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotype of every individual.\nl1::Integer: the first focal loci.\nl2::Integer: the second focal loci.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_average_linkage_diseq-Tuple{Vector{<:Matrix{<:Integer}}, NamedTuple}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_average_linkage_diseq","text":"calc_average_linkage_diseq(genotypes::Vector{<:Matrix{<:Integer}}, loci::NamedTuple)\n\nCompute the linkage disequilibrium between each loci and return a table of values of the  average correlation between loci of each trait.\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotype of every individual.\nloci::NamedTuple: the name and loci range of each trait of interest.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_trait_correlation-Tuple{Vector{<:Matrix{<:Integer}}, Union{UnitRange{<:Integer}, Vector{<:Integer}}, Union{UnitRange{<:Integer}, Vector{<:Integer}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_trait_correlation","text":"calc_trait_correlation(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci_range1::Union{UnitRange{<:Integer},Vector{<:Integer}},\n    loci_range2::Union{UnitRange{<:Integer},Vector{<:Integer}}\n)\n\nCompute the correlation between two traits.\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotype of every individual.\nloci_range1::Union{UnitRange{<:Integer},Vector{<:Integer}}: the loci range of the first trait.\nloci_range2::Union{UnitRange{<:Integer},Vector{<:Integer}}: the loci range of the second trait.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_all_trait_correlations-Tuple{Vector{<:Matrix{<:Integer}}, NamedTuple}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_all_trait_correlations","text":"calc_all_trait_correlations(genotypes::Vector{<:Matrix{<:Integer}}, loci::NamedTuple)\n\nCompute the correlation between each trait.\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotype of every individual.\nloci::NamedTuple: the name and loci range of each trait of interest.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.sort_y-Tuple{Vector{<:Real}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.sort_y","text":"sort_y(y_locations::Vector{<:Real})\n\nSort the y coordinate indices into 10 vectors [0, 0.1), [0.1, 0.2), etc.\n\nExample\n\njldoctest sort_y([0.01, 0.5, 0.24, 0.9]) 10-element Vector{<:Vector{Int64}}:  [1]  []  [3]  []  []  [2]  []  []  []  [4]\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.sort_locations-Tuple{AbstractArray{<:Real}, Real}","page":"DataAnalysis","title":"HZAM.DataAnalysis.sort_locations","text":"sort_locations(A::AbstractArray{<:Real}, bin_size::Real)\n\nSort the given list of numbers into bins of a given size spanning 0 to 1.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.average_gene_data-Tuple{Vector{<:NamedTuple}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.average_gene_data","text":"average_gene_data(gene_data::Vector{<:NamedTuple})\n\nCompute the mean for each field of the given list of composite datatypes. \n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_traits_additive-Tuple{Vector{<:Matrix{<:Integer}}, Union{UnitRange{<:Integer}, Vector{<:Integer}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_traits_additive","text":"calc_traits_additive(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::Union{UnitRange{<:Integer},Vector{<:Integer}}\n)::Vector{Float32}\n\nCompute the mean values of the genotypes passed to it at the given loci. Used to determine  trait values in an additive way.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.average_data_per_phenotype-Tuple{Vector{<:Integer}, Vector{<:Matrix{<:Integer}}, Union{UnitRange{<:Integer}, Vector{<:Integer}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.average_data_per_phenotype","text":"average_data_per_phenotype(\n    data::Vector{<:Integer},\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::Union{UnitRange{<:Integer},Vector{<:Integer}}\n)\n\nCompute the average number of mates for males of each phenotype for a given  loci range.\n\nArguments\n\ndata::Vector{<:Integer}: a list of the data to be averaged for each element in the genotypes vector\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotypes for each individual \nloci::Union{UnitRange{<:Integer},Vector{<:Integer}: the loci range governing the trait of interest.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.count_phenotypes_at_loci-Tuple{Vector{<:Matrix{<:Integer}}, Union{UnitRange{<:Integer}, Vector{<:Integer}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.count_phenotypes_at_loci","text":"function count_phenotypes_at_loci(\n    genotypes::Vector{<:Matrix{<:Integer}},\n    loci::Union{UnitRange{<:Integer},Vector{<:Integer}}\n)\n\nReturn the number of individuals with each phenotype for the given loci range.\n\nArguments\n\ngenotypes::Vector{<:Matrix{<:Integer}}: the genotypes of every individual\nloci::Union{UnitRange{<:Integer},Vector{<:Integer}}: the loci range over which the phenotypes are calculated.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.find_fixed_alleles-Tuple{Vector{<:Matrix{<:Integer}}}","page":"DataAnalysis","title":"HZAM.DataAnalysis.find_fixed_alleles","text":"find_fixed_alleles(genotypes::Vector{<:Matrix{<:Integer}})\n\nDetermine at which loci the population has lost its genetic diversity.\n\n\n\n\n\n","category":"method"},{"location":"data_analysis/#HZAM.DataAnalysis.calc_chi_squared-Tuple{Vector{<:Matrix{<:Integer}}, Integer}","page":"DataAnalysis","title":"HZAM.DataAnalysis.calc_chi_squared","text":"calc_chi_squared(genotypes::Vector{<:Matrix{<:Integer}}, locus::Integer)\n\nCompare the observed genotype frequencies at the given locus to those expected from  Hardy-Weinberg analysis.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM","page":"HZAM","title":"HZAM","text":"","category":"section"},{"location":"#Contents","page":"HZAM","title":"Contents","text":"","category":"section"},{"location":"","page":"HZAM","title":"HZAM","text":"","category":"page"},{"location":"#Functions","page":"HZAM","title":"Functions","text":"","category":"section"},{"location":"","page":"HZAM","title":"HZAM","text":"HZAM.run_one_HZAM_sim(\n        w_hyb::Real, \n        S_AM::Real, \n        ecolDiff::Real, \n        intrinsic_R::Real;\n    )\nHZAM.run_HZAM_set(set_name::String, intrinsic_R::Real, ecolDiff::Real;)\nHZAM.plot_output_field(\n    outcomes::Array{:Real},\n    sim_params::Array{<:HZAM.DataAnalysis.SimParams}\n)\nHZAM.load_from_folder(dir::String)\nHZAM.load_from_csv(filepath::String)\nHZAM.convert_to_CSV(\n    outcome_array::Array{<:HZAM.DataAnalysis.OutputData},\n    field_name::Symbol,\n    output_folder::String\n)\nHZAM.plot_fitnesses(fitnesses::Vector{<:Dict})\nHZAM.plot_population_tracking_data(filepath::String)\nHZAM.summarize_gene_correlations(dir::String)\nHZAM.plot_phenotypes(phenotypes;)","category":"page"},{"location":"#HZAM.run_one_HZAM_sim-NTuple{4, Real}","page":"HZAM","title":"HZAM.run_one_HZAM_sim","text":"run_one_HZAM_sim(\n    w_hyb::Real, \n    S_AM::Real, \n    ecolDiff::Real, \n    intrinsic_R::Real; \n    <keyword arguments>\n)\n\nRun a single HZAM simulation.\n\nArguments\n\nw_hyb::Real: the hybrid fitness.\nS_AM::Real: the strength of assortative mating.\necolDiff::Real: the ecological difference between the two species.\nintrinsic_R::Real: the intrinsic growth rate.\nK_total::Integer=20000: the carrying capacity of the environment.\nmax_generations::Integer=1000: the number of generations that the simulation will run for.\ntotal_loci::Integer=6: the total number of loci in the genome.\nfemale_mating_trait_loci=1:3: the loci specifying the female's mate preference.\nmale_mating_trait_loci=1:3: the loci specifying the male's mating trait.\ncompetition_trait_loci=1:3: the loci specifying the ecological trait (used in fitness related to resource use).\nhybrid_survival_loci=1:3: the loci specifying the probability of survival to adulthood.\nsurvival_fitness_method:String=\"epistasis\": the method used to calculate the probability of survival to adulthood.\nper_reject_cost=0: the fitness loss of female per male rejected (due to search time, etc.). Can take values of 0 to 1.\nsigma_disp=0.05: the standard deviation of the normal distribution determining how far offspring will disperse from their mothers.\nsigma_comp=0.01: the standard deviation for the normal curve used in calculating local density.\ndo_plot=true: the program will display a plot of the locations and hybrid indices of every individual while the simulation is running if this is true.\nplot_int=10: the interval (measured in generations) between updating the plot.\ngene_plot=false: if true, generates phenotype plots.\nsave_plot=false: if true, saves each plot to a PNG file.\ntrack_spatial_data=false: if true, keeps track of bimodality, gene flow, variance, cline widths, and cline positions.\ntrack_population_data=true: if true, keeps track of population size, hybridness, overlap, and hybrid zone width.\ntrack_fitness=false: if true, keeps track of the average number of offspring for each phenotype.\ntrack_mating_success=false: if true, keeps track of the average number of mates per male for each phenotype.\ntrack_phenotypes=false: if true, keeps track of the number of individuals with each phenotype.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.run_HZAM_set-Tuple{String, Real, Real}","page":"HZAM","title":"HZAM.run_HZAM_set","text":"run_one_HZAM_sim(\n    w_hyb::Real, \n    S_AM::Real, \n    ecolDiff::Real, \n    intrinsic_R::Real; \n    <keyword arguments>\n)\n\nRun the simulation for 64 combinations of hybrid fitness and assortative mating strength and  store the outcome of each simulation in a JLD2 file.\n\nArguments\n\nset_name::String: the name assigned to the set of simulations.\nintrinsic_R::Real: the intrinsic growth rate.\necolDiff::Real: the ecological difference between the two species. Between 0 and 1.\nK_total::Integer=20000: the carrying capacity of the environment.\nmax_generations::Integer=1000: the number of generations that the simulation will run for.\ntotal_loci::Integer=6: the total number of loci in the genome.\nfemale_mating_trait_loci=1:3: the loci specifying the female's mate preference.\nmale_mating_trait_loci=1:3: the loci specifying the male's mating trait.\ncompetition_trait_loci=1:3: the loci specifying the ecological trait (used in fitness related to resource use).\nhybrid_survival_loci=1:3: the loci specifying the probability of survival to adulthood.\nsurvival_fitness_method:String=\"epistasis\": the method used to calculate the probability of survival to adulthood.\nper_reject_cost=0: the fitness loss of female per male rejected (due to search time, etc.). Can take values of 0 to 1.\nsigma_disp=0.05: the standard deviation of the normal distribution determining how far offspring will disperse from their mothers.\ntrack_spatial_data=false: if true, keeps track of bimodality, gene flow, variance, cline widths, and cline positions.\ntrack_population_data=true: if true, keeps track of population size, hybridness, overlap, and hybrid zone width.\ntrack_fitness=false: if true, keeps track of the average number of offspring for each phenotype.\ntrack_mating_success=false: if true, keeps track of the average number of mates per male for each phenotype.\ntrack_phenotypes=false: if true, keeps track of the number of individuals with each phenotype.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.plot_output_field-Tuple{Array{:Real}, Array{<:HZAM.DataAnalysis.SimParams}}","page":"HZAM","title":"HZAM.plot_output_field","text":"plot_output_field(\n    outcomes::Array{:Real},\n    sim_params::Array{<:DataAnalysis.SimParams}\n)\n\nCreate a heatmap of an output variable vs hybrid fitness and assortative mating.\n\nArguments\n\noutcomes::Array{:Real}: the output from the simulation to be displayed.\nsim_params::Array{<:DataAnalysis.SimParams}: the simulation parameters resulting in the outcomes.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.load_from_folder-Tuple{String}","page":"HZAM","title":"HZAM.load_from_folder","text":"load_from_folder(dir::String)\n\nLoad the data from each simulation output file stored in the given directory into an array  organized by hybrid fitness and assortative mating strength.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.load_from_csv-Tuple{String}","page":"HZAM","title":"HZAM.load_from_csv","text":"load_from_csv(filepath::String)\n\nRead the data from a CSV file and return vectors for hybrid fitness, assortative mating  strength, and the output variable.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.convert_to_CSV-Tuple{Array{<:HZAM.DataAnalysis.OutputData}, Symbol, String}","page":"HZAM","title":"HZAM.convert_to_CSV","text":"convert_to_CSV(\n    outcome_array::Array{<:DataAnalysis.OutputData},\n    field_name::Symbol,\n    output_folder::String\n)\n\nConvert an array of outcomes to a csv file for the given output field.\n\nArguments\n\noutcome_array::Array{<:DataAnalysis.OutputData}: the array of output data from a set of simulations.\nfield_name::Symbol: the name of the output field of interest.\noutput_folder::String: the destination for the CSV.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.plot_fitnesses-Tuple{Vector{<:Dict}}","page":"HZAM","title":"HZAM.plot_fitnesses","text":"plot_fitnesses(fitnesses::Vector{<:Dict})\n\nProduce a plot of fitnesses per phenotype over time.\n\nArguments\n\nfitnesses::Vector{<:Dict}: number of offspring per phenotype each generation.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.plot_population_tracking_data-Tuple{String}","page":"HZAM","title":"HZAM.plot_population_tracking_data","text":"plot_population_tracking_data(filepath::String)\n\nCreate plots of the population size, hybrid zone width, hybrid index, and population overlap vs time.\n\nArguments\n\nfilepath::String:: the filepath for the file containing the outcome of the simulation.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.summarize_gene_correlations-Tuple{String}","page":"HZAM","title":"HZAM.summarize_gene_correlations","text":"summarize_gene_correlations(dir::String)\n\nCreate plots of the gene correlations between different traits vs SAM and whyb for  different combinations of the same loci controlling different traits.\n\nArguments\n\ndir::String: the folder where the outcomes of all the simulations are stored.\n\n\n\n\n\n","category":"method"},{"location":"#HZAM.plot_phenotypes-Tuple{Any}","page":"HZAM","title":"HZAM.plot_phenotypes","text":"plot_phenotypes(phenotypes; filepath=\"phenotype_frequencies.png\")\n\nCreate plots of mating preference, mating cue, and hybrid survival trait phenotype  frequencies over time.\n\nArguments\n\nfilepath=\"phenotype_frequencies.png\": the destination filepath for the plot image.\n\n\n\n\n\n","category":"method"},{"location":"#Index","page":"HZAM","title":"Index","text":"","category":"section"},{"location":"","page":"HZAM","title":"HZAM","text":"","category":"page"}]
}
